### PR TITLE
Add mode test 

### PR DESCRIPTION
**Pull request by CongMa13**
Add support for any modes of tensors

For example, modes could be 
```
A: K5K3M0M5K4K1M3K2M1K0M4M2 
B: M4M0M3M5K4M2K2K3M1K5K1K0 
C: K1K4K3M2M4K5M5M0K0M1M3K2
```

One change to support the mode unit test is sequence test parameters. 

Combined test parameter is the only type supported before. But in mode test we cannot combine lengths with modes. 

for example, length [1, 2], [1, 2, 3] cannot combine with modes [0, 1], [0, 1, 2]

So added support to sequence test parameters. 

Example: test.yaml

```
Param1: 2.1
Param2: ALGO_DEFAULT
Param3: HIP_C_32F 
Param4:
  - 1
  - 2
  - 3
 Param5:
  - a
  - b
  - c
 ```

The test params are 
```
2.1, ALGO_DEFAULT, HIP_C_32F, 1, a
2.1, ALGO_DEFAULT, HIP_C_32F, 2, b
2.1, ALGO_DEFAULT, HIP_C_32F, 3, c
```

image

### CODE REVIEW

File: library/src/contraction/contraction_solution.cpp [hunk 1]
```diff
@@ -24,10 +24,140 @@
...
+        std::vector<std::size_t> normal_b_ns_ks_strides(MaxNumDimsK + MaxNumDimsN, 1);
+        std::vector<int32_t>     normal_b_ns_ks_modes(MaxNumDimsK + MaxNumDimsN, -1);
+        std::vector<std::size_t> normal_e_ms_ns_lengths(MaxNumDimsM + MaxNumDimsN, 1);
+        std::vector<std::size_t> normal_e_ms_ns_strides(MaxNumDimsM + MaxNumDimsN, 1);
+        std::vector<int32_t>     normal_e_ms_ns_modes(MaxNumDimsM + MaxNumDimsN, -1);
+        int                      mOffset = 0;
+        int                      nOffset = 0;
+
+        // reorder m, n in A, B
+        for(int i = 0; i < e_ms_ns_modes.size(); i++)
```
Comment (cgmillette) [inline library/src/contraction/contraction_solution.cpp:58]: Is it possible to explain each of these steps a bit more? And possibly with a sample of the case they cover?

**Review by cgmillette — DISMISSED** [review]
Looks good - just maybe needs some more description on the normalization procedure and why that's needed 👍🏼

File: library/src/contraction/contraction_solution.cpp (hunk 1 above)
Comment (CongMa13) [reply library/src/contraction/contraction_solution.cpp:58]: image

Example:
|Matrix | mode | length| stride |
|-|-|-|-|
|A | [3, 0, 4, 1] | [ 2, 3, 4, 5] | [60, 20, 5, 1] |
|B | [4, 2, 3] | [ 4, 6, 2] | [12, 2, 1] |
|E | [2, 1, 0] | [6, 5, 3] | [ 15, 3, 1]|

**find out M,N,K**  
    mode_E ∩ mode_A == M
    mode_E ∩ mode_B == N
    mode_A ∩ mode_B == K

And we hold the order of M and N in E, order of K in A


|Matrix | mode | 
|-|-|
|A | [K0, M1, K1, M0]
|B | [K1, N0, K0] | 
|E | [N0, M0, M1] | 

**We need to permute and extend the matrices to have**

|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5, K0, K1, K2, K3, K4, K5] |
|normalized  B | [N0, N1, N2, N3, N4, N5, K0, K1, K2, K3, K4, K5] | 

**Line 60 - 79 change `normalized A` and `normalized B` to** 
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1] |
|normalized  B | [N0] | 

**Line 81 - 91 extend `normalized A` and `normalized B` to** 

extended dimension: length is 1 and stride is the same as M1 and N0's stride.
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5] |
|normalized  B | [N0, N1, N2, N3, N4, N5] | 

**Line 94 - 110 add K  to `normalized A` and `normalized B`** 
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5, K0, K1] |
|normalized  B | [N0, N1, N2, N3, N4, N5, K0, K1] | 

**Line 112 - 122 extend K in `normalized A` and `normalized B`** 
|Matrix | mode |  length| stride |
|-|-|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5, K0, K1, K2, K3, K4, K5] | [5, 3, 1, 1, 1, 1, 2, 4, 1, 1, 1, 1] | [1, 20, 20, 20, 20, 20, 60, 5, 5, 5, 5 ,5] |
|normalized  B | [N0, N1, N2, N3, N4, N5, K0, K1, K2, K3, K4, K5] | [4, 1, 1, 1, 1, 1, 2, 4, 1, 1, 1, 1] | [2, 2, 2, 2, 2, 2, 1, 12, 12, 12, 12, 12] |

**Line 125 - 133 generate `normalized E`** 

Copy modes from `normalized A` and `normalized B`
|Matrix | mode |   length| stride |
|-|-|-|-|
|E | [N0, M0, M1] | [6, 5, 3] | [ 15, 3, 1]|
|normalized E | [M0, M1, M2, M3, M4, M5, N0, N1, N2, N3, N4, N5] | [ 5, 3, 1, 1, 1, 1, 6, 1, 1, 1, 1, 1] | [3, 1, 1, 1, 1, 1, 15, 15, 15, 15, 15, 15] |

**Line 133 - 148 permute the E by switching length and strides**

**Review by mkarunan — DISMISSED** [review]
Looks great

**Review by cgmillette — APPROVED** [review]

**Pull request merged**